Public/Src/Cache/ContentStore/Distributed/NuCache/RocksDbContentLocationDatabase.cs

```diff
@@ -507,6 +529,197 @@ private byte[] GetKey(ShortHash hash)
+                                return AddOrGetContentHashListResult.SinglePhaseMixingError;
+                            }
+
+                            // Replace if incoming has better determinism or some content for the existing entry is missing.
+                            if (oldContentHashList == null || oldDeterminism.ShouldBeReplacedWith(determinism) ||
+                                        !IsContentAvailable(context, oldContentHashList))
+                            {
+                                // TODO(jubayard): SQLite impl runs this with exclusive access to the DB to avoid data races. We have no way to do this here.
+                                store.Put(key, SerializeContentHashList(contentHashListWithDeterminism), nameof(Columns.Metadata));
+
+                                // Accept the value
+                                return new AddOrGetContentHashListResult(CacheMiss(determinism));
+                            }
+
+                            // If we didn't accept the new value because it is the same as before, just with a not
+                            // necessarily better determinism, then let the user know.
+                            if (oldContentHashList != null && oldContentHashList.Equals(contentHashList))
+                            {
+                                return new AddOrGetContentHashListResult(CacheMiss(oldDeterminism));
+                            }
+
+                            // If we didn't accept a deterministic tool's data, then we're in an inconsistent state
+                            if (determinism.IsDeterministicTool)
+                            {
+                                return new AddOrGetContentHashListResult(
+                                    AddOrGetContentHashListResult.ResultCode.InvalidToolDeterminismError,
+                                    oldContentHashListWithDeterminism.ContentHashListWithDeterminism);
+                            }
+
+                            // If we did not accept the given value, return the value in the cache
+                            return new AddOrGetContentHashListResult(oldContentHashListWithDeterminism);
+                        });
+
+                    // Success for this status here may actually be an error that's been returned
+                    return status.Succeeded ? status.Result : new AddOrGetContentHashListResult(status.Failure.CreateException());
+                }, Counters[ContentLocationDatabaseCounters.AddOrGetContentHashList]);
+        }
+        
+        private ContentHashListWithDeterminism CacheMiss(CacheDeterminism determinism)
+        {
+            return new ContentHashListWithDeterminism(null, determinism);
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<StructResult<StrongFingerprint>> EnumerateStrongFingerprints(OperationContext context)
+        {
+            var result = new List<StructResult<StrongFingerprint>>();

```

Comment:
Minor: Can you use yield here to return an enumerator without getting the whole list first? This is probably minor because constructing the whole list is probably quite fast.

Reply:
That was my initial response as well. It is doable, but more work without any clear benefit. There are two ways to do it:

1. Implement an iterator that holds the key value store lock throughout until nothing more can be retrieved from the prefix search. This is potentially dangerous, because we have no idea what the user is doing with the enumeration, so potential locks could cause issues with other parts of the system.
2. An iterator that keeps taking and releasing the DB lock when necessary. This may be problematic as well because the iterator could fail in a strange ways. A quick example: if a task is iterating, then goes dark, and before it tries to fetch a new element the database is restored from a different place, we will close it while the iteration is still going on and hence the next fetch for the enumerable may fail in a weird manner.

There doesn't seem to be any good solution, so I just bit the bullet.

---
In reply to: [302767227](https://github.com/microsoft/BuildXL/pull/571#discussion_r302767227) [](ancestors = 302767227)